sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CollectionsShouldImplementGenericInterface.cs

```diff
@@ -0,0 +1,106 @@
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static readonly Dictionary<string, KnownType> nongenericToGenericMapping =
+            new Dictionary<string, KnownType>
+            {
+                [KnownType.System_Collections_ICollection.TypeName] = KnownType.System_Collections_Generic_ICollection_T,
+                [KnownType.System_Collections_IList.TypeName] = KnownType.System_Collections_Generic_IList_T,
+                [KnownType.System_Collections_IEnumerable.TypeName] = KnownType.System_Collections_Generic_IEnumerable_T,
+                [KnownType.System_Collections_CollectionBase.TypeName] = KnownType.System_Collections_ObjectModel_Collection_T,
+            };
+
+        private static readonly ISet<KnownType> genericTypes = nongenericToGenericMapping.Values.ToHashSet();
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+                {
+                    var classDeclaration = c.Node as ClassDeclarationSyntax;
+
+                    var implementedTypes = classDeclaration?.BaseList?.Types;
+                    if (implementedTypes == null)
+                    {
+                        return;
+                    }
+
+                    var issues = new List<Diagnostic>();
+                    foreach (var typeSyntax in implementedTypes)
+                    {
+                        var typeSymbol = c.SemanticModel.GetSymbolInfo(typeSyntax.Type).Symbol?.GetSymbolType();
+                        if (typeSymbol == null)
+                        {
+                            continue;
+                        }
+
+                        if (typeSymbol.OriginalDefinition.IsAny(genericTypes))
+                        {
+                            return;
+                        }
+
+                        var suggestedGenericType = SuggestGenericCollectionType(typeSymbol);
+                        if (suggestedGenericType != null)
+                        {
+                            issues.Add(Diagnostic.Create(rule,
+                                        classDeclaration.Identifier.GetLocation(),
+                                        suggestedGenericType));
+                        }
+                    }
+
+                    issues.ForEach(c.ReportDiagnostic);
+                },
+                SyntaxKind.ClassDeclaration);
+        }
+
+        private static string SuggestGenericCollectionType(ITypeSymbol typeSymbol)
+        {
+            KnownType suggestedGenericType;
+            if (nongenericToGenericMapping.TryGetValue(typeSymbol.ToDisplayString(), out suggestedGenericType))
```

Comment:
If you use ImmutableDictionary you will have the GetValueOrDefault extension method and this could be written as:

    return nongenericToGenericMapping.GetValueOrDefault(typeSymbol.ToDisplayString())?.TypeName;

perhaps we should some day add such method for normal IDictionary<K,V>